### ISSUE TITLE
Add algorithm attribute to histogram field mapper.

### ISSUE THREAD

**Issue by martijnvg** [elastic/elasticsearch#108208]
In Elasticsearch there are two histogram algorithms available: tdigest and hdr. The histogram field mapper is algorithm agnostic. A tdigest or hdr histogram can be stored in a field mapped as histogram. However the features using the histogram field like aggregations don't know what kind of histogram is stored in a histogram field. So it is possible to run a percentile aggregation configured with `tdigest` on a histogram field that stores a hdr histogram. The result that gets produced will not be useful at all. Ideally this shouldn't be allowed. But today this validation doesn't occur because the field mapper doesn't record what algorithm is used. 

Additionally downsampling needs to know what algorithm is used, otherwise downsampling can't merge histograms. This is the reason why today downsampling doesn't support metric fields of type histogram.

The proposal is to add an algorithm attribute to the histogram field type. So that other features can make use of that. I don't think we can make this a required argument for bwc reasons from the start (but maybe many versions from now). 

Relates to #74213

**Comment by axw** [comment]
> The histogram field mapper is algorithm agnostic.

Should we also consider revisiting this decision? Having explicitly algorithm-specific histogram field mappers would also solve the problems described, and would also enable more efficient storage. The HDR Histogram field mapper would not need to store the bucket values, but could use the algorithm parameters & offsets to derive the bucket values.

Similar to HDR Histogram, see https://github.com/elastic/elasticsearch/pull/104298, a PoC I implemented of a mapper & aggs based on the OpenTelemetry exponential histogram metric type: https://opentelemetry.io/docs/specs/otel/metrics/data-model/#exponentialhistogram

**Comment by martijnvg** [comment]
> Should we also consider revisiting this decision? Having explicitly algorithm-specific histogram field mappers would also solve the problems described, and would also enable more efficient storage.

Yes, I agree and that is what this issue is about. We either change the current histogram field type to be (optionally) algorithm-specific, or we can introduce new algorithm-specific histogram field types.

**Comment by jdcryans** [comment]
This issue is superseded by https://github.com/elastic/elasticsearch/issues/128622
